_drafts/hw_actors.md

```diff
@@ -0,0 +1,150 @@
+
+The approach is to use this model and map priorities to interrupt vectors with user-defined priority. Note that this model contains no threads and runs at a single shared stack. It is inherently asynchronous and even applicable to multiprocessor systems since (in the case of GIC) you may trigger interrupts on other CPUs too.
+
+
+## The concept
+
+
+Since the actor model itself is a very complex topic for a single pocket article, instead of talking about abstractions, I mostly focus on their representation in the MCU environment.
+
+In the classic actor model, an actor is just a message handler similar to interrupt handlers. Unlike threads, it has no dedicated stack and its purpose is to react to messages arriving into its internal mailbox.
+
+During message handling, the actor may:
+- create new actors
+- send messages to other actors
+- read/write its local state
+
+Despite the actor and its mailbox are tightly coupled in this model, I think it is less practical than relying on separate queues. The latter case is more flexible since an actor may choose which queue to poll depending on the message that is being processed.
+
+Since there are no threads, the main() purpose is just an initialization. It ends with an infinite loop waiting for interrupts. Note that after initialization completes, there are no other  executable entities except interrupt handlers.
+
+Interrupts are split into two classes: interrupts producing messages and interrupts consuming messages. Also, there are only three types of software objects: queues, actors and messages.
+
+It is expected that interrupts designated to peripheral devices will post messages to queues which in turn have subscribed actors, so, pseudocode for the interrupt handler may be written as:
+
+
+        device_interrupt_handler() {
+            message = alloc_message(pool)
+            ... read device registers and set message payload ...
+            push_to_queue(queue, message)
+        }
+
+
+Interrupt handlers that are dedicated to actors contain just a call to function, examining a list of active actors with priority corresponding to that vector:
+
+
+        actor_interrupt_handler() {
+            schedule_actors(VECTOR_ID)
+        }
+
+
+On initialization phase, each actor subscribes to some queue. Once message arrives to that queue the actor is activated. Since each actor may be subscribed to exactly one queue, subscription may be implemented as return value: the actor should return a valid pointer to the queue it wants to be subscribed to. Return values may change during an actor’s lifetime but must always be a valid queue. This is because of the absence of internal queue: when an actor has not been subscribed to a queue, it will have no chance to run again. So, pseudocode for an actor is shown below:
+
+
+        actor(self, message) {
+            ... message handling ...
+            free(message)
+            return queue_to_be_subscribed_to
+        }
```

Comment:
I think it would be nice to have a working code sample for this. Could you add something to the `example` directory?